### DISCUSSION TITLE
Questions RE Unsafe Accessors

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#92568] [Q&A] [answered]
I don't have a .net 8.0 environment so I can't test this out but I have a few questions about this:

1.  Do parameter names have to match?
2.  If I have a mismatch and try to access a private member that does not exist, is it a runtime or compile time error?
3.  Is this zero overhead?
4.  Does this potentially enable scenarios like: ThirdPartyLibrary.SetIsLicensed(true)
?

**Comment by huoyaoyuan** [comment]
> 4\. Does this potentially enable scenarios like: ThirdPartyLibrary.SetIsLicensed(true)
>     ?

It can, but it does not *enable*. You can always do such things in reflection.

> 3\. Is this zero overhead?

Yes, JIT specially recognizes them and translates to direct access as the members were public.

**Comment by paulbartrum** [comment]
> 1. Do parameter names have to match?

No. But the parameter types do have to match exactly.

> 2. If I have a mismatch and try to access a private member that does not exist, is it a runtime or compile time error?

It's a runtime error. The member resolution is done at JIT-time, not compile-time.

Note that the expected usage is via source generator, where provided you code it correctly you'll always generate the exact required method name, parameter types and modifiers.

**Comment by TonyValenti** [comment]
If I have a private method named "set_Value" and I want the function named "Value_Set" is there an attribute I can apply to include the "real" name of the method?

**Reply by paulbartrum** [reply, marked as answer]
You mean like this? The name of the private method is specified using the attribute.
```c#
[UnsafeAccessor(UnsafeAccessorKind.Method, Name = "set_Value")]
extern static void Value_Set(ClassWithSetValueMethod c, object value);
```
